### PR TITLE
support for bipartite hypergraph representation

### PR DESCRIPTION
**Pull request by pszufe**
Initial implementation for bipartite hypergraph representation

### CODE REVIEW

**Comment by bkamins** [comment]
Asignee oznacza, że ten ktoś ma to robić; review to oddzielna rola

File: docs/src/index.md [hunk 1]
```diff
@@ -1,3 +1,5 @@
 # SimpleHypergraphs.jl
 
 Documentation for SimpleHypergraphs.jl
+
+For details please go to the *reference* section.
```
Comment (bkamins) [inline docs/src/index.md]: add link

File: docs/src/reference.md [hunk 2]
```diff
@@ -43,6 +43,9 @@ Hypergraph info
 size(::Hypergraph)
 getvertices(::Hypergraph, ::Int)
 gethyperedges(::Hypergraph, ::Int)
+
+SimpleHypergraphs.BipartiteView
```
Comment (bkamins) [inline docs/src/reference.md]: we export it so I guess we do not have to qualify this?

File: src/SimpleHypergraphs.jl [hunk 3]
```diff
@@ -1,8 +1,13 @@
 module SimpleHypergraphs
 
+using LightGraphs
+
 export Hypergraph, getvertices, gethyperedges, hg_load, hg_save
+export add_vertex!, add_hyperedge!
```
Comment (bkamins) [inline src/SimpleHypergraphs.jl]: why these two are not referenced in reference.md?

File: src/bipartite.jl [hunk 4]
```diff
@@ -0,0 +1,96 @@
+"""
+    BipartiteView{T<:Real} <: AbstractGraph{Int64}
+
```
Comment (bkamins) [inline src/bipartite.jl]: unnecessary newline

File: src/bipartite.jl [hunk 5]
```diff
@@ -0,0 +1,96 @@
+"""
+    BipartiteView{T<:Real} <: AbstractGraph{Int64}
```
Comment (bkamins) [inline src/bipartite.jl]: I guess `Int` not `Int64`?

File: src/bipartite.jl [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+
+Creates a bipartite view of a hypergraph.
```
Comment (bkamins) [inline src/bipartite.jl]: Better to say "Create".

File: src/SimpleHypergraphs.jl [hunk 7]
```diff
@@ -1,8 +1,13 @@
...
+export BipartiteView, shortest_path
 
 include("types.jl")
+include("bipartite.jl")
```
Comment (bkamins) [inline src/SimpleHypergraphs.jl]: typically it is better to keep types in one file as it is easier to manage code later. but we can go the way you propose, then `types.jl` file should be renamed.

File: src/bipartite.jl [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+Note this is a view - changes to the original hypergraph will be automatically reflected in the view.
+
+**Constructors**
```
Comment (bkamins) [inline src/bipartite.jl]: This format is discouraged (it later floods REPL). See e.g. how `Array` is documented. OTOH Examples section is useful to have.

File: src/bipartite.jl [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+Construct a bipartite view of a hypergraph for processing with LightGraphs
+
+
+**Arguments**
+
+* `h` : a hypergraph
+
+"""
+struct BipartiteView{T<:Real} <: AbstractGraph{Int64}
+    v2he::Vector{Dict{Int,T}}
```
Comment (bkamins) [inline src/bipartite.jl]: why not simply hold a `Hypergraph`. This breaks encapsulation and will be harder to maintain in the long range.

File: src/bipartite.jl [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    he2v::Vector{Dict{Int,T}}
+    function BipartiteView{T}(h::Hypergraph{T}) where {T<:Real}
+        new(h.v2he,h.he2v)
+    end
+end
+
+
+LightGraphs.nv(b::BipartiteView) =
```
Comment (bkamins) [inline src/bipartite.jl]: I would make one-liners of this and also below without separating newline.

File: src/bipartite.jl [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+     length(b.v2he)+length(b.he2v)
+
+LightGraphs.vertices(b::BipartiteView) =
+     Base.OneTo(nv(b))
+
+
+LightGraphs.ne(b::BipartiteView) =
+    sum(length.(b.v2he))+sum(length.(b.he2v))
```
Comment (bkamins) [inline src/bipartite.jl]: I do not understand why you make these two sums. Especially as they are equal in value.

File: src/bipartite.jl [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+      n1 .+ keys(b.v2he[v])
+    else
+      keys(b.he2v[v-n1])
+    end
+end
+
+function LightGraphs.has_edge(b::BipartiteView,s,d)
+    n1 = length(b.v2he)
+    if s <= n1
+        if d <= n1 return false; end
```
Comment (bkamins) [inline src/bipartite.jl]: `d > n1 && has_key(b.v2he[s], d - n1)`

File: src/bipartite.jl [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        has_key(b.v2he[s], d-n1)
+    else
+        if d > n1 return false; end
```
Comment (bkamins) [inline src/bipartite.jl]: `d <= n1 && has_key(b.v2he[s], d - n1)`

File: src/bipartite.jl [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        has_key(b.he2v[s-n1], d)
+    end
+end
+
+
+LightGraphs.outneighbors(b::BipartiteView,v::Integer) =
+    LightGraphs.all_neighbors(b::BipartiteView,v)
```
Comment (bkamins) [inline src/bipartite.jl]: here and everywhere space after `,`

File: src/bipartite.jl [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+Construct a bipartite view of a hypergraph for processing with LightGraphs
+
+
+**Arguments**
+
+* `h` : a hypergraph
+
+"""
+struct BipartiteView{T<:Real} <: AbstractGraph{Int64}
```
Comment (bkamins) [inline src/bipartite.jl]: Probably `Int` not `Int64`?

File: src/bipartite.jl [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+LightGraphs.outneighbors(b::BipartiteView,v::Integer) =
+    LightGraphs.all_neighbors(b::BipartiteView,v)
+
+LightGraphs.inneighbors(b::BipartiteView,v::Integer) =
+    LightGraphs.all_neighbors(b::BipartiteView,v)
+
+
+function LightGraphs.SimpleGraph(b::BipartiteView)
+    g = SimpleGraph(length(b.v2he)+length(b.he2v))
```
Comment (bkamins) [inline src/bipartite.jl]: why not `nv(b)`?

File: src/bipartite.jl [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+end
+
+
+
+LightGraphs.is_directed(b::BipartiteView) = false
+
+"""
+    shortest_path(g::BipartiteView,source, target)
+
+Finds the shortest path between two vertices
```
Comment (bkamins) [inline src/bipartite.jl]: the same comment as for `BipartiteView`

File: src/bipartite.jl [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+**Arguments**
+
+* `b` : a bipartite view of a hypergraph
+* `source` : a starting vertex
+* `target` : a destination vertex
+"""
+function shortest_path(g::BipartiteView,source::Int, target::Int)
+    dj = dijkstra_shortest_paths(g, source)
+    res = enumerate_paths(dj)[target]
+    filter(n -> (n <= length(g.v2he)),res)
```
Comment (bkamins) [inline src/bipartite.jl]: what if `target` or `source` is an edge in hypergraph. You support it, but then filter it out.

Comment (bkamins) [inline src/bipartite.jl]: actually you can take every second vertex which should be faster.

File: src/bipartite.jl [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+
+**Arguments**
+
+* `b` : a bipartite view of a hypergraph
+* `source` : a starting vertex
+* `target` : a destination vertex
+"""
+function shortest_path(g::BipartiteView,source::Int, target::Int)
+    dj = dijkstra_shortest_paths(g, source)
+    res = enumerate_paths(dj)[target]
```
Comment (bkamins) [inline src/bipartite.jl]: is res a single path or a collection of paths? (I have not checked)

File: src/types.jl [hunk 20]
```diff
@@ -130,3 +130,13 @@ Returns hyperedges for a given vertex
 * `v_id` : an identifier of a vertex
 """
 @inline gethyperedges(h::Hypergraph, v_id::Int) = h.v2he[v_id]
+
+function add_vertex!(h::Hypergraph{T};hyperedges::Dict{Int,T} = Dict{Int,T}()) where T <: Real
+    push!(h.v2he,hyperedges)
+    length(h.v2he)
+end
+
+function add_hyperedge!(h::Hypergraph{T};vertices::Dict{Int,T} = Dict{Int,T}()) where T <: Real
```
Comment (bkamins) [inline src/types.jl]: documentation missing.
probably we also should then support:
1. adding an existing vertex to an existing hyperedge
2. removing methods (a vertex, a hyperedge, a vertex from a hyperedge)

File: src/types.jl [hunk 21]
```diff
@@ -130,3 +130,13 @@ Returns hyperedges for a given vertex
 * `v_id` : an identifier of a vertex
 """
 @inline gethyperedges(h::Hypergraph, v_id::Int) = h.v2he[v_id]
+
+function add_vertex!(h::Hypergraph{T};hyperedges::Dict{Int,T} = Dict{Int,T}()) where T <: Real
+    push!(h.v2he,hyperedges)
```
Comment (bkamins) [inline src/types.jl]: this method and also one below is wrong as it updates only one collection while both collections vertices->hyperedges and hyperedges->vertices should be updated in both cases.

**Comment by bkamins** [comment]
Also if you add functions for mutating a hypergraph we should add a function that checks if it is not corrupted

File: src/bipartite.jl (hunk 5 above)
Comment (pszufe) [reply src/bipartite.jl]: changed

File: src/bipartite.jl (hunk 4 above)
Comment (pszufe) [reply src/bipartite.jl]: changed. maybe let's edit such things directly in the code.

File: docs/src/index.md (hunk 1 above)
Comment (pszufe) [reply docs/src/index.md]: added

File: src/bipartite.jl (hunk 19 above)
Comment (pszufe) [reply src/bipartite.jl]: edited method description.
Seems that LightGraphs implementation returns only one path (to be verified)

**Comment by bkamins** [comment]
OK to remove branch?

**Pull request merged**